File: Add Code Here/JAVA/Sorting Leetcode Questions/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
      Arrays.sort(nums);
      List<List<Integer>> result = new ArrayList<>();
      for (int i = 0; i < nums.length && nums[i] <= 0; i++) {
          
        if (i == 0 || nums[i - 1] != nums[i]) {
            
          Set<Integer> set = new HashSet<>();
          for (int j = i + 1; j < nums.length; j++) {
              
            int target = -1 * (nums[i] + nums[j]);
            if (set.contains(target)) {
                
              result.add(Arrays.asList(nums[i], nums[j], target));
              while (j + 1 < nums.length && nums[j] == nums[j + 1]) {
                j++;
              }
            }
            set.add(nums[j]);
          }
        }
      }
      return result;
    }
  }

File: Add Code Here/C++/leetcode/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        if(n==0)
            return {};
        sort(nums.begin(), nums.end());
        vector<vector<int>>v;
        int k=nums[0];
        for(int i=0;i<n-2;i++)
        {
            if(i==0 || k!=nums[i])
            {
                cout<<i<<" "<<k<<" ";
                k=nums[i];
                cout<<k<<endl;
                int l=i+1, h = n-1;
                while(l<h)
                {
                    if(nums[l]+nums[h]==-k)
                    {
                        v.push_back({k,nums[l],nums[h]});
                        while(l<h && nums[l]==nums[l+1]) l++;
                        while(l<h && nums[h]==nums[h-1]) h--;
                        l++;
                        h--;
                    }
                    else if(nums[l]+nums[h]>-k)
                        h--;
                    else l++;
                }
            }
        }
        return v;
    }
};
